### PR TITLE
feat: register autoload function if not using composer

### LINKED ISSUE #638: Vendor autoload.php doesn't exist?
I'm getting the following error:

> Warning: require(/xxxxxxxx/sendgrid/vendor/autoload.php): failed to open stream: No such file or directory in /xxxxxxxx/sendgrid/sendgrid-php.php on line 14

I found these related issues but there's no proper solution offered there: https://github.com/sendgrid/sendgrid-php/issues/407

Why is this vendor not file not included in github??
The [PHP example you guys provide](https://sendgrid.com/docs/Integrate/Code_Examples/v3_Mail/php.html) doesn't even work this way.

### PR DESCRIPTION
**Pull request by EvaLok**
fixes #638

based on the documentation, the library should work without using composer, but doesn't. this PR registers an autoload function based on the directory structure of the library.

### CODE REVIEW

**Comment by EvaLok** [comment]
@thinkingserious it looks like the tests failed because of server load or something; may want to rerun them during off hours?

File: sendgrid-php.php [hunk 1]
```diff
@@ -11,10 +11,33 @@
...
+                $path = str_replace('\\', '/', $class);
+
+                $pieces = explode('/', $path);
+                array_shift($pieces); // we don't use the first piece (SendGrid)
+                $file = array_pop($pieces);
+
+                $dirPath = __DIR__ . '/lib/' . strtolower(implode('/', $pieces));
+                $filePath = $dirPath . '/' . $file . '.php';
+
+                if ( file_exists($filePath) ) {
```
Comment (eshanholtz) [inline sendgrid-php.php:32]: nit: remove spaces around conditional statement in parenthesis for style consistency

File: sendgrid-php.php [hunk 2]
```diff
@@ -11,10 +11,33 @@
...
+                    require_once($filePath);
+                }
+
+                if ( ! class_exists($class) ) {
```
Comment (eshanholtz) [inline sendgrid-php.php:36]: nit: remove spaces around conditional statement and `!` operator for style consistency.

**Comment by thinkingserious** [comment]
Hello @EvaLok,

I was able to confirm that the library does work without using composer. Here are the steps I used:

`wget https://github.com/sendgrid/sendgrid-php/releases/download/7.9.1/sendgrid-php.zip`
`unzip sendgrid-php.zip`
`touch temp.php`

Here are the contents of `temp.php`:

```php
<?php
require_once 'sendgrid-php.php';

$email = new \SendGrid\Mail\Mail();
$email->setFrom("example@example.com", "Example User");
$email->setSubject("Sending with Twilio SendGrid is Fun");
$email->addTo("example@example.com", "Example User");
$email->addContent("text/plain", "and easy to do anywhere, even with PHP");
$email->addContent(
    "text/html", "<strong>and easy to do anywhere, even with PHP</strong>"
);
$sendgrid = new \SendGrid(getenv('SENDGRID_API_KEY'));
try {
    $response = $sendgrid->send($email);
    print $response->statusCode() . "\n";
    print_r($response->headers());
    print $response->body() . "\n";
} catch (Exception $e) {
    echo 'Caught exception: '. $e->getMessage() ."\n";
}
```

`php temp.php`

The email was sent successfully. 

Please advise on how you came to the conclusion that this library does not work without composer. Thank you!

With best regards,

Elmer

**Comment by EvaLok** [comment]
@thinkingserious using either a git clone or a zip of the source code will cause the error. looking at the zip file you used, it seems like you packaged extra files (some composer generated stuff) which are not included in the repository.. and which would of course explain why it worked for you and not for me. i based my assumption of the feature not working at all on this comment https://github.com/sendgrid/sendgrid-php/issues/638#issuecomment-576412237 which wasn't answered, so i assumed that there was no working solution and thus didn't think to try the non source code package. in any case, the solution i submitted will allow the library to work without the extra package files, but whether or not you wish to use it is of course up to you.

**Comment by eshanholtz** [comment]
@EvaLok 

I like the idea of having a non-composer solution to autoload. If you can address the few pieces of style feedback I gave, I'd like to merge this work. 

Thanks!

**Comment by EvaLok** [comment]
i'm not really familiar with your company's coding styles/standards, but i
can recommend integrating a linter into your automated test / build process
in order to avoid this class of issue altogether (eg:
https://cs.symfony.com/) =)

**Comment by EvaLok** [comment]
oh - i thought i should add: it is mentioned in your documentation, but of
course the library still depends on `sendgrid/php-http-client` having been
installed (but only for certain classes) - that is a potential gotcha for
anyone using the standalone library (i don't recall off the top of my head
whether or not the packaged version includes this dependency in the
`vendor` directory, but i assume that it does). the reason i got into this
mess in the first place was that i was working on a legacy
dependency-locked project and thus couldn't use composer. honestly if it's
working with the packaged version, then i think it might be better to stick
with what's proven to be working unless you want to reconfigure all of your
automated testing to also check standalone functionality =)

**Comment by eshanholtz** [comment]
That's a good point. I'm going to go ahead and close this PR since we have a working solution for not using Composer.

**Pull request closed without merging**